### ISSUE TITLE
User defined cleanReplacements should be executed before defaults.

### ISSUE THREAD

**Issue by mgilson** [yabwe/medium-editor#925]
In the paste extension, the user defined `cleanReplacements` are appended to the medium editor default replacements (https://github.com/yabwe/medium-editor/blob/master/src/js/extensions/paste.js#L143).

As a user, I believe that I should have the opportunity to transform the markup _before_ medium transforms it into a state where I can no longer recognize the nodes I'm looking for.

If you agree, I think it should be as easy as changing

```javascript
replacements = createReplacements().concat(this.cleanReplacements || []);
```

to

```javascript
replacements = (this.cleanReplacements || []).concat(createReplacements());
```

**Comment by nmielnik** [comment]
@mgilson So I agree it should be possible, but I don't agree that we should change the current code to pre-pend instead of append the user defined `cleanReplacements` because of backward compatibility issues, and the fact that this has worked for users for a while.  

If we were to make this possible, I would rather add an additional option (something like `preCleanReplacements` or a better name) that would allow users to pass in options that run both before (`preCleanReplacements`) and after (`cleanReplacements`).

If you have a specific need for this, you can always override the paste extension, and change just that chunk of code, but we would also gladly accept a PR to add this new option :smile:

**Comment by mgilson** [comment]
I can probably do something like that when I get a little time.

With a name like `preCleanReplacements`, I'd also expect there to be a `postCleanReplacements` (maybe an alias for `cleanReplacements` for backward compatibility?)  Does that sound like a good idea to you?

--------

My actual reason for requesting this was because the paste extension doesn't handle underline text pasted from google docs.  I came up with a hack:

```
var monkeyPatchPasteExtension = function(editor) {
  var pasteExt = editor.getExtensionByName('paste');
  pasteExt['cleanupSpans'] = cleanupSpans;
};


/**
 * Cleans up the spans which have replace-with classes.
 * @param {!Element} containerEl
 * @return {boolean|undefined} [description]
 * @this {!MediumEditor.Extension}
 */
var cleanupSpans = function(containerEl) {
  var i,
      el,
      newEl,
      innerHtml,
      spans = containerEl.querySelectorAll('.replace-with'),
      isCEF = function(el) {
        return (el && el.nodeName !== '#text' &&
            el.getAttribute('contenteditable') === 'false');
      };

  var replaceClasses = ['bold', 'italic', 'underline'];
  var tagNames = {'bold': 'b', 'italic': 'i', 'underline': 'u'};
  for (i = 0; i < spans.length; i += 1) {
    el = spans[i];
    innerHtml = el.innerHTML;
    replaceClasses.forEach(function(className) {
      if (el.classList.contains(className)) {
        var tagName = tagNames[className];
        innerHtml = '<' + tagName + '>' + innerHtml + '</' + tagName + '>';
      }
    });
    newEl = this.document.createElement('div');
    newEl.innerHTML = innerHtml;
    newEl = newEl.firstChild;

    el.parentNode.replaceChild(newEl, el);
  }

  spans = containerEl.querySelectorAll('span');
  for (i = 0; i < spans.length; i += 1) {
    el = spans[i];

    // bail if span is in contenteditable = false
    if (MediumEditor.util.traverseUp(el, isCEF)) {
      return false;
    }

    // remove empty spans, replace others with their contents
    MediumEditor.util.unwrap(el, this.document);
  }
};
```

This monkey-patching is obviously ugly, but I couldn't come up with a better way to do it without forking the component entirely. . .
This works for the following "cleanReplacements":

```javascript
...
cleanReplacements: [
    new RegExp(/<span[^>]*text-decoration:\s*underline[^>]*>/gi),
    '<span class="replace-with underline">',
]
```

But it stops working if the text is bold and underlined because the medium-editor defined replacements clobber the underline elements before my regex has a chance to see them.  Obviously, we get combinatorial explosion if you try to support _everything_, so I'm not suggesting that medium-editor start supporting underline as well (Oh no, the next user wants to preserve strike-through text!!! ahh!!!).

But allowing my regex to run before yours would make it so that I can get my hack to work...

And, having said all this, I guess I'm wondering if there are more options that could make this even more flexible in the future.  Personally, I would _love_ a solution that doesn't stick me with [parsing HTML with Regex](http://stackoverflow.com/a/1732454/748858) -- But I'm not sure that exists easily.

Anyway, I think I've left the beaten path.  To bring it back on track, I think I can implement the `preCleanReplacements` option you described.  If you'd like, I can make a `postCleanReplacements` alias for `cleanReplacements` (but I won't do that if you don't want it) and we can certainly try to think of better names for these options (though I haven't come up with anything while I've been typing this).

**Comment by nmielnik** [comment]
This was fixed by #940
